libp2p/crypto/ecnist.nim

```diff
@@ -232,37 +237,46 @@ proc random*(t: typedesc[EcPrivateKey], kind: EcCurveKind): EcPrivateKey =
   var seeder = brPrngSeederSystem(nil)
   brHmacDrbgInit(addr rng, addr sha256Vtable, nil, 0)
   if seeder(addr rng.vtable) == 0:
-    raise newException(ValueError, "Could not seed RNG")
+    return err(EcRngError)
   var ecimp = brEcGetDefault()
-  result = new EcPrivateKey
-  result.buffer = newSeq[byte](BR_EC_KBUF_PRIV_MAX_SIZE)
+  var res = new EcPrivateKey
+  res.buffer = newSeq[byte](BR_EC_KBUF_PRIV_MAX_SIZE)
   if brEcKeygen(addr rng.vtable, ecimp,
-                addr result.key, addr result.buffer[0],
+                addr res.key, addr res.buffer[0],
                 cast[cint](kind)) == 0:
-    raise newException(ValueError, "Could not generate private key")
+    err(EcKeyGenError)
+  else:
+    ok(res)
 
-proc getKey*(seckey: EcPrivateKey): EcPublicKey =
+proc getKey*(seckey: EcPrivateKey): EcResult[EcPublicKey] =
   ## Calculate and return EC public key from private key ``seckey``.
-  doAssert(not isNil(seckey))
+  if isNil(seckey):
+    return err(EcKeyIncorrectError)
+
   var ecimp = brEcGetDefault()
   if seckey.key.curve in EcSupportedCurvesCint:
     var length = getPublicKeyLength(cast[EcCurveKind](seckey.key.curve))
-    result = new EcPublicKey
-    result.buffer = newSeq[byte](length)
-    if brEcComputePublicKey(ecimp, addr result.key,
-                            addr result.buffer[0], unsafeAddr seckey.key) == 0:
-      raise newException(EcKeyIncorrectError, "Could not calculate public key")
+    var res = new EcPublicKey
+    res.buffer = newSeq[byte](length)
+    if brEcComputePublicKey(ecimp, addr res.key,
+                            addr res.buffer[0], unsafeAddr seckey.key) == 0:
+      err(EcKeyIncorrectError)
+    else:
+      ok(res)
   else:
-    raise newException(EcKeyIncorrectError, "Incorrect private key")
+    err(EcKeyIncorrectError)
```

Comment:
It is better to have here different error, `EcKeyUnsupportedCurveError` or something like that, otherwise both parts returns equal errors and it hard to understand what happens.